### PR TITLE
New serverless pattern - s3-lambda-comprehend-dynamodb-batch-sentiment-analysis

### PR DESCRIPTION
**Pull request by awsarippa**
*Issue #, if available:*

*Description of changes:*
The PR is raised to showcase a serverless pattern to perform batch sentiment analysis with services like S3, Lambda, Comprehend, and DynamoDB.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: s3-lambda-comprehend-cdk-python/CODE_OF_CONDUCT.md
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/CODE_OF_CONDUCT.md:1]: This file can be deleted as it is not necessary for each individual pattern

File: s3-lambda-comprehend-cdk-python/CONTRIBUTING.md
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/CONTRIBUTING.md:1]: Can be deleted as well

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+   - architecture.png
+   - user_input_text.txt
+- requirements.txt
+- source.bat
+```
+
+## Common Errors & Troubleshooting
+
+### "ValueError: Must setup local AWS configuration with a region supported by AWS Services."
+Solution: You must set an aws region with `export AWS_DEFAULT_REGION=<your-region>`
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/DEVELOPMENT.md:18]:
```suggestion
Solution: You must set an AWS region with `export AWS_DEFAULT_REGION=<your-region>`
```

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+### Error creating role
+```
+botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the CreateRole operation: User: <user-arn> is not authorized to perform: iam:CreateRole on resource: <role-arn> because no identity-based policy allows the iam:CreateRole action
+```
+Solution: you must ensure the Iam role you are using has sufficient permissions to create Iam roles
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/DEVELOPMENT.md:24]:
```suggestion
Solution: you must ensure the IAM role you are using has sufficient permissions to create IAM roles
```

File: s3-lambda-comprehend-cdk-python/LICENSE
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/LICENSE:1]: Not necessary

File: s3-lambda-comprehend-cdk-python/README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+## Architecture
+![Diagram](src/architecture.jpg)
+
+### What resources will be created?
+This CDK code will create the following:
+   - 1 S3 bucket (to hold the text for which sentiment analysis has to be done)
+   - 1 Lambda (to invoke the Comprehend API)
+   - 1 DynamoDB (to store the result of Sentiment Analysis)
+   - 1 IAM role (for the Lambda to invoke Comprehend and DynamoDB)
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:12]: * please use "one" instead of "1"
* please use Lambda function instead of Lambda

File: s3-lambda-comprehend-cdk-python/README.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+and deploy with
+```
+cdk deploy
+```
+
+The deployment will create a S3 bucket, a Lambda and a S3 bucket.
+
+## How it works
+The S3 bucket acts as a placeholder to upload the text, required for performing Sentiment Analysis. In the demonstration, we use the contents inside `src\user_input_text.txt`. 
+The file is uploaded to S3 bucket, which triggers the Lambda. The Lambda processes the contents within the uploaded file and invokes the Comprehend's BatchDetectSentiment API and the analyzed response from Comprehend service is stored in DynamoDB.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:69]:
```suggestion
The file is uploaded to S3 bucket, which triggers the Lambda function. The Lambda function processes the contents within the uploaded file and invokes the Comprehend's `BatchDetectSentiment` API and the analyzed response from Comprehend service is stored in DynamoDB.
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+## Testing
+Upon successful deployment of the stack, the Output section would provide the `S3 Bucket` in the CDK environment. 
+Upload the sample file `src\user_input_text.txt` into the `S3 bucket`. The upload action triggers the `Lambda` and the input text is analyzed.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:73]:
```suggestion
Upload the sample file `src\user_input_text.txt` into the `S3 bucket`. The upload action triggers the Lambda function and the input text is analyzed.
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+A response as below is returned from the `Comprehend` service and stored in a `DynamoDB` with file name as the `partition key`.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:74]:
```suggestion
A response as below is returned from Amazon Comprehend and stored in a DynamoDB table with file name as the partition key. 
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+## How it works
+The S3 bucket acts as a placeholder to upload the text, required for performing Sentiment Analysis. In the demonstration, we use the contents inside `src\user_input_text.txt`. 
+The file is uploaded to S3 bucket, which triggers the Lambda. The Lambda processes the contents within the uploaded file and invokes the Comprehend's BatchDetectSentiment API and the analyzed response from Comprehend service is stored in DynamoDB.
+
+## Testing
+Upon successful deployment of the stack, the Output section would provide the `S3 Bucket` in the CDK environment. 
+Upload the sample file `src\user_input_text.txt` into the `S3 bucket`. The upload action triggers the `Lambda` and the input text is analyzed.
+A response as below is returned from the `Comprehend` service and stored in a `DynamoDB` with file name as the `partition key`. 
+As per specific use-cases, the responses can be converted into a flat file to be stored in `S3`.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:75]:
```suggestion
As per specific use-cases, the responses can be converted into a flat file to be stored in S3.
```

File: s3-lambda-comprehend-cdk-python/app.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        )
+
+        # lambda function for processing the incoming request from S3 bucket
+        lambda_function = lambda_.Function(
+            self,
+            "DetectSentimentLambda",
+            runtime=lambda_.Runtime.PYTHON_3_10,
+            handler="lambda_function.lambda_handler",
+            code=lambda_.Code.from_asset(os.path.join(DIRNAME, "src")),
+            timeout=Duration.minutes(10),
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/app.py:68]: Is a timeout of 10 minutes really necessary?

File: s3-lambda-comprehend-cdk-python/app.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            memory_size=2048,
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/app.py:69]: That seems like a lot of memory, why is that necessary?

File: s3-lambda-comprehend-cdk-python/README.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+   - 1 Lambda (to invoke the Comprehend API)
+   - 1 DynamoDB (to store the result of Sentiment Analysis)
+   - 1 IAM role (for the Lambda to invoke Comprehend and DynamoDB)
+
+## Requirements
+
+### Development Environment
+**Cloud 9**
+
+This code for this pattern is executed in AWS Cloud9 environment. The EC2 instance used is m5.large (8 GiB RAM + 2 vCPU).
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:19]: Why do I need a Cloud9 instance for this? Is it not possible to run this locally as well?

File: s3-lambda-comprehend-cdk-python/README.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+### AWS setup
+**Region**
+
+If you have not yet run `aws configure` and set a default region, you must do so, or you can also run `export AWS_DEFAULT_REGION=<your-region>`. The region used in the demonstration is us-east-1 as Bedrock service is available only in limited regions.
+
+**Authorization**
+
+You must use a role that has sufficient permissions to create Iam roles, as well as cloudformation resources
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:28]:
```suggestion
You must use a role that has sufficient permissions to create IAM roles, as well as CloudFormation resources
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+### Development Environment
+**Cloud 9**
+
+This code for this pattern is executed in AWS Cloud9 environment. The EC2 instance used is m5.large (8 GiB RAM + 2 vCPU).
+
+### AWS setup
+**Region**
+
+If you have not yet run `aws configure` and set a default region, you must do so, or you can also run `export AWS_DEFAULT_REGION=<your-region>`. The region used in the demonstration is us-east-1 as Bedrock service is available only in limited regions.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:24]:
```suggestion
If you have not yet run `aws configure` and set a default region, you must do so, or you can also run `export AWS_DEFAULT_REGION=<your-region>`. The region used in the demonstration is us-east-1 as Amazon Bedrock is currently not available in every region.
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+**Authorization**
+
+You must use a role that has sufficient permissions to create Iam roles, as well as cloudformation resources
+
+#### Python >=3.8
+Make sure you have [python3](https://www.python.org/downloads/) installed at a version >=3.8.x in the CDK environment. The demonstration has used python version 3.10.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:31]:
```suggestion
Make sure you have [python3](https://www.python.org/downloads/) installed at a version >=3.8.x in the CDK environment. The demonstration uses python 3.10.
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+Install the required dependencies (aws-cdk-lib and constructs) into your Python environment 
+```
+pip install -r requirements.txt
+```
+
+### Gather and deploy resources with the CDK
+
+First synthesize, which executes the application, defines which resources will be created, and translates this into a cloudformation template
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:52]: What do you mean by "executes the application"?

Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:52]:
```suggestion
First synthesize, which executes the application, defines which resources will be created, and translates this into a CloudFormation template
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+```
+cdk synth
+```
+Now bootstrap, which provisions the resources you'll use when deploying the application
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:56]: This is a one-time process and only necessary before deploying your first CDK application

File: s3-lambda-comprehend-cdk-python/README.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+```
+cdk bootstrap
+```
+and deploy with
+```
+cdk deploy
+```
+
+The deployment will create a S3 bucket, a Lambda and a S3 bucket.
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:65]: Please use Lambda function and remove the duplicate S3 bucket

File: s3-lambda-comprehend-cdk-python/app.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+DIRNAME = os.path.dirname(__file__)
+
+
+class S3LambdaComprehendServerless(Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+
+        self.user_input_bucket = s3.Bucket(
+            self,
+            "s3-upload-input-text",
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/app.py:32]: S3 bucket names are globally unique. This will lead to errors as the bucket already exists. Consider making this a parameter or instructing the user to change it before deployment.

File: s3-lambda-comprehend-cdk-python/app.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+                "environment": "dev",
+                "dynamodb_table": dynamodb_table.table_name,
+            },
+        )
+
+        # lambda version
+        version = lambda_function.current_version
+
+        # lambda alias
+        alias = lambda_.Alias(self, "Alias", alias_name="Prod", version=version)
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/app.py:80]: Why is an alias needed here? I don't see it used in the rest of the file

File: s3-lambda-comprehend-cdk-python/CONTRIBUTING.md
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/CONTRIBUTING.md:1]: Done; removed the file.

File: s3-lambda-comprehend-cdk-python/CODE_OF_CONDUCT.md
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/CODE_OF_CONDUCT.md:1]: Done; removed.

File: s3-lambda-comprehend-cdk-python/README.md (hunk 3 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md]: Done; updated README.md

File: s3-lambda-comprehend-cdk-python/README.md (hunk 14 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md]: The intention was to help under what the `cdk synth` command does here. 
I've tried to summarize the contents from the AWS documentation here. 

https://docs.aws.amazon.com/cdk/v2/guide/hello_world.html#hello_world_tutorial_synth

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+```
+s3-lambda-comprehend-cdk-python
+- app.py
+- cdk.json
+- src/
+   - lambda_function.py
+   - architecture.png
+   - user_input_text.txt
+- requirements.txt
+- source.bat
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/DEVELOPMENT.md:12]: where is this file or can that line be deleted?

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+## Common Errors & Troubleshooting
+
+### "ValueError: Must setup local AWS configuration with a region supported by AWS Services."
+Solution: You must set an AWS region with `export AWS_DEFAULT_REGION=<your-region>`
+
+### Error creating role
+```
+botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the CreateRole operation: User: <user-arn> is not authorized to perform: iam:CreateRole on resource: <role-arn> because no identity-based policy allows the iam:CreateRole action
+```
+Solution: you must ensure the IAM role you are using has sufficient permissions to create Iam roles
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/DEVELOPMENT.md:24]:
```suggestion
Solution: you must ensure the IAM role you are using has sufficient permissions to create IAM roles
```

File: s3-lambda-comprehend-cdk-python/README.md [hunk 21]
```diff
@@ -0,0 +1,127 @@
+# S3 -> Lambda -> Comprehend
+This patterns shows CDK deployment on how to leverage Amazon S3, AWS Lambda, Amazon Comprehend and Amazon DynamoDB to perform Batch Sentiment Analysis in a serverless fashion.
+
+## Architecture
+![Diagram](src/architecture.jpg)
+
+### What resources will be created?
+This CDK code will create the following:
+   - One S3 bucket (to hold the text for which sentiment analysis has to be done)
+   - One Lambda (to invoke the Comprehend API)
```
Comment (bfreiberg) [inline s3-lambda-comprehend-cdk-python/README.md:10]:
```suggestion
   - One Lambda function (to invoke the Comprehend API)
```

**Review by bfreiberg — COMMENTED** [review]
Thanks for the quick update. I've commented a couple more minor things

File: s3-lambda-comprehend-cdk-python/README.md (hunk 15 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md]: Updated the messaging.

File: s3-lambda-comprehend-cdk-python/app.py (hunk 17 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/app.py:33]: Updated app.py with a comment asking user to replace the bucket name with a globally unique name.

File: s3-lambda-comprehend-cdk-python/app.py (hunk 8 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/app.py]: That was a mistake. Intention was to keep one minute/60 seconds.

File: s3-lambda-comprehend-cdk-python/app.py (hunk 9 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/app.py]: Reduced the memory size to 512.

File: s3-lambda-comprehend-cdk-python/app.py (hunk 18 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/app.py]: Removed alias, as it is not required at this point of time.

File: s3-lambda-comprehend-cdk-python/README.md (hunk 10 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md:19]: The idea is to use Cloud9 to perform the CDK deployment. To use locally, user may have to use AWS SAM CLI to locally test the AWS CDK applications. 
I've updated the messaging.

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md (hunk 1 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/DEVELOPMENT.md]: Updated the comment.

File: s3-lambda-comprehend-cdk-python/DEVELOPMENT.md (hunk 19 above)
Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/DEVELOPMENT.md]: Line removed.

File: s3-lambda-comprehend-cdk-python/README.md (hunk 10 above)
Comment (bfreiberg) [reply s3-lambda-comprehend-cdk-python/README.md]: Sorry, maybe there is some confusion on my side. Could you reiterate why a Cloud9 environment is necessary and a local executing of `cdk synth && cdk deploy` is not advised? I've managed to deploy it from my local machine just fine. And how does the AWS SAM CLI relate to this?

Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md]: Apologies, my intention was to keep everything within AWS ecosystem (which includes the CDK synth and deploy). There is nothing blocking users from using local environment. It's just that I used a Cloud9 environment as per the convenience.

Users can also use AWS SAM CLI to deploy CDK apps. 
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-cdk-testing.html

Comment (bfreiberg) [reply s3-lambda-comprehend-cdk-python/README.md]: Ok, thanks for the clarifications. I think it would be best to add a line about the possibility to do it locally as well. The Cloud9 instance will incur cost (even if minimal) and users might forget to delete it afterwards. On that note, I don't see a callout to terminate the instance at the end in the README.

Comment (awsarippa) [reply s3-lambda-comprehend-cdk-python/README.md]: Thanks a lot for the observation.
Yes, I've added a comment showcasing capability to execute CDK deployment from local environment and towards the end to remind users to terminate the Cloud9 EC2 instance.

**Comment by bfreiberg** [comment]
Looks good, thank you for your contribution. The PR will be merged soon

**Comment by awsarippa** [comment]
@bfreiberg - Thanks a lot Ben for your time and effort the review and suggestions. Appreciate the same.

**Comment by bls20AWS** [comment]
will go live on https://serverlessland.com/patterns/s3-lambda-comprehend-cdk-python within 24 hours
